Add support for all streams minimal

theViz343 opened the pull request:
<!-- See README for documentation, or ask in #zulip-terminal if unclear -->
### What does this PR do, and why?
This PR is very similar to #1408 and #1419 and combines elements of both. This PR keeps the initial commits from #1419 but replaces the general accessor methods with only an attribute-specific necessary method i.e "name".


### Outstanding aspect(s)    <!-- DELETE SECTION IF EMPTY -->
<!-- In what ways is this not fully implemented/functioning? Compared to a discussion/issue? -->
<!-- Do you not understand something? Are you unsure about a certain approach? Want feedback? -->
- [x] Could add more attribute accessor methods (like "color")

### External discussion & connections
<!-- [x] all that apply, specifying topic and adding numbers after # for issues/PRs -->
- [x] Discussed in **#zulip-terminal** in [`Adding streams throws key error`](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/Adding.20streams.20throws.20key.20error)
- [x] Fully fixes #816 
- [ ] Partially fixes issue #
- [x] Builds upon previous unmerged work in PR #1419 and PR #1408 
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [x] Merge will enable work on #1387 

### How did you test this?
<!-- [x] all that apply -->
- [ ] Manually - Behavioral changes
- [ ] Manually - Visual changes
- [x] Adapting existing automated tests
- [x] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [x] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [x] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [x] It has a commit summary describing the  motivation and reasoning for the change
- [x] It individually passes linting and tests
- [x] It contains test additions for any new behavior
- [x] It flows clearly from a previous branch commit, and/or prepares for the next commit

### Visual changes    <!-- DELETE SECTION IF NO VISUAL CHANGE -->
<!-- Zulip tips at https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html -->
<!-- For video, try asciinema; after uploading, embed using
[![yourtitle](https://asciinema.org/a/<id>.png)](https://asciinema.org/a/<id>)
-->
<!-- NOTE: Attached videos/images will be clearer from smaller terminal windows -->

theViz343 committed:
model: Add stream and subscription id helper methods.

This commit adds three helper methods - _get_stream_from_id,
_get_subscription_from_id, and get_all_stream_ids. These three
helper methods help in preparation for adding stream and
subscription property accessor methods in the next commits.

Tests and fixture added.

theViz343 committed:
model: Improve typing of stream_dict.

With the revamp of the Subscription TypedDict in an earlier commit, we
use it to improve the typing of stream_dict in this commit.

The `is_old_stream` is removed from stream fixtures since ZFL 14
removes them from the Streams object. This field was always
equivalent to stream_weekly_traffic != null on the same object.

Tests and fixtures updated to support the new typing.

theViz343 committed:
api_types: Update Stream and Subscription typeddict fields.

This commit updates the `date_created` and
`stream_post_policy` fields present in the Stream typeddict,
and the `role` field in the Subscription typeddict.

For `date_created`, since servers with ZFL<30 do not include
the field, but ZT does add it with a value of None to ensure
consistency, the type is changed to
`NotRequired[Optional[int]]`.

For `stream_post_policy`, the field is only included after
Zulip 3.0 / ZFL 1, so it has been changed to a NotRequired
field.

For `role`, the field has been removed in Zulip 6.0 (ZFL 133),
so it has been removed from the Subscriptions typeddict and
also from the unsubscribed_streams_fixture.

theViz343 committed:
model/boxes/buttons/views: Add stream_name_from_id method to model.

This commit introduces the stream_name_from_id method in
order to limit the direct accessing of stream_dict outside the
model. The commit also replaces the access of the name attribute
from stream_dict with the new method.

Tests added and updated.

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -891,6 +891,7 @@ def test_keypress_NEXT_UNREAD_TOPIC_stream(
 
         mid_col_view.model.stream_dict = {1: {"name": "stream"}}
         mid_col_view.model.next_unread_topic_from_message_id.return_value = (1, "topic")
+        mid_col_view.model.get_stream_name_from_id.return_value = "stream"
```
It's important to adjust the tests, not just to pass, but to represent what is now necessary. 

In this commit you're using the function instead of the stream_dict for the name - so is it now necessary to set the stream_dict in the test?

Many of the test changes need to be checked for this.

neiljp commented on zulipterminal/model.py:
```diff
@@ -1297,6 +1297,10 @@ def _get_all_stream_ids(self) -> List[int]:
         id_list.extend(stream_id for stream_id in self._never_subscribed_streams)
         return id_list
 
+    def get_stream_name_from_id(self, stream_id: int) -> str:
```
Based on the existing `stream_id_from_name`, let's drop the `get_`.

neiljp commented on zulipterminal/api_types.py:
```diff
@@ -244,9 +250,11 @@ class Subscription(Stream):
     color: str
 
+    # Deprecated fields
+    # role: int  # Removed in Zulip 6.0 (feature level 133)
```
I think I noted somewhere that it's worth including here that it wasn't actively used by us (or the web app even?), and introduced at some other ZFL?

neiljp commented on tests/core/test_core.py:
```diff
@@ -124,18 +125,22 @@ def test_narrow_to_stream(
         controller.view.message_view = mocker.patch("urwid.ListBox")
         controller.model.stream_dict = {
-            stream_id: {
+            stream_id: general_stream,
+        }
+        controller.model.stream_dict[stream_id].update(
+            {
                 "color": "#ffffff",
                 "name": stream_name,
             }
-        }
+        )
```
It's good that this commit focuses on the stream/subscription migration, but
- it doesn't seem that the color is important to the test(s)?
- that could make this transition smaller/simpler
- can we replace the stream_dict code by setting the return_value in the model, in the later commits? (if we don't already)

theViz343 replied:
For the third point, which function's return_value are you referring to?

neiljp replied:
I meant that where you're moving to use the functions that access the name and color, instead of using stream_dict, can we now mock those functions and set the return value directly?

Essentially, we want to ensure we move away from the stream_dict fixture (or equivalent manual dicts), in addition to directly accessing the model attribute.

neiljp reviewed (commented):
@theViz343 It's good to see another go at this one, though there are some comments from before to keep up with, including:
- https://github.com/zulip/zulip-terminal/pull/1419/commits/29fe057222efcfddacbcd6cb3b4fab0b40c33bd1#r1280046556
- https://github.com/zulip/zulip-terminal/pull/1419/commits/1ced8ec95217618b3e8f4e658728dfc4cf97d8f0#r1280070702

I'm happy with the core of this, but other than the points above was initially interested to see if the early new data in the Model you added could be confirmed in test_init using new fixtures. To see that you added fixtures which had similar data to test later code confirmed that would be useful, to test the data flow through the Model from the initial_data into the accessors.

theViz343 commented on zulipterminal/model.py:
```diff
@@ -1281,6 +1281,22 @@ def _register_non_subscribed_streams(
             stream["stream_id"]: stream for stream in never_subscribed_streams
         }
 
+    def _get_stream_from_id(self, stream_id: int) -> Union[Subscription, Stream]:
```
If we do that, should I cast the `subscribed` and `unsubscribed` streams returned by this method to `Stream`? That might make the usage of this method clearer too, ie. to only access stream properties (and not subscription properties) from the returned stream.

neiljp replied:
A cast shouldn't be necessary once you have the two different functions - you don't have that in the latest version of the code you just pushed?

theViz343 replied:
I've casted the stream_dict and the unsubscribed_streams to a `Stream` in the `_get_stream_from_id` in the latest commit, so that methods which use the `_get_stream_from_id` always access only the stream properties and not the subscription properties (which would then warrant the use of `_get_subscription_from_id`).

neiljp replied:
Ah right, but I don't remember seeing an explicit `cast()` to work around mypy - I assume this works in the type system since a Stream is a subset of a Subscription.

theViz343 commented on tests/conftest.py:
```diff
@@ -243,21 +244,25 @@ def general_stream() -> Dict[str, Any]:
         "audible_notifications": False,
         "description": "General Stream",
         "rendered_description": "General Stream",
-        "is_old_stream": True,
```
I could not find any documentation in the api docs or CZO discussion for this field. What was the field's original purpose, if you know? I can then add it's reason for removal in the commit text with more context.

neiljp replied:
It's in the API changelog (https://zulip.com/api/changelog), and I found it on czo to be zulip/zulip#15409.

theViz343 committed:
model/boxes/messages: Add subscription_color_from_id method to model.

This commit introduces the subscription_color_from_id method in
order to limit the direct accessing of stream_dict outside
the model. The commit also replaces the access of the color
attribute from stream_dict with the new method.

Tests added and updated.

neiljp commented on tests/core/test_core.py:
```diff
@@ -124,18 +125,17 @@ def test_narrow_to_stream(
-            }
+            stream_id: general_stream,
         }
+        controller.model.stream_dict[stream_id]["name"] = stream_name
```
If it's necessary to keep this line for the test, can the stream_dict be removed instead later, when we switch to the new function?

neiljp reviewed (commented):
@theViz343 Thanks for the update again :+1:

Overall this is good, and the commits are generally clean and well tested.

I think most of my non-minor feedback has come about since this refactoring is complex - tidying types, correcting fixtures, fixing inconsistencies, adding data, using new accessors, removing old fixtures, integrating with existing code, updating old accessors - and so there are a lot of potential rough edges, some of which only surface later.

The stream_dict removals in the tests could be left for another time, certainly the complex cases.

I think clarifying the stream/subscription validity functions are likely the standout issue to address here, and which we need to be a little careful about since they're already in use, and we don't interact with all streams in every case.

theViz343 committed:
refactor: Reduce the direct usage of stream dict in tests.

This commit removes the direct use of stream dicts in most
of the tests and if necessary, patches the stream accessor
methods created in a previous commit.

theViz343 committed:
api_types: Refactor Subscription TypedDict into Stream and Subscription.

This commit refactors the Subscription typeddict to separate Stream
and Subscription TypedDicts. The Subscription TypedDict now
inherits the Stream TypedDict. This is a preparatory refactor to
later introduce the unsubscribed_streams and
never_subscribed_streams to the model.

theViz343 committed:
model/api_types: Add data structures to support non-subscribed streams.

This commit creates data structures, similar to stream_dict, to
support unsubscribed and never-subscribed streams. These streams
are available in the fetched initial data which is used to populate
the new data structures using the _register_non_subscribed_streams
function.

Test updated and fixtures added.
